pkg/gofr/datasource/opentsdb/opentsdb.go

```diff
@@ -0,0 +1,418 @@
+	DetectDeltaNum int
+
+	// The maximal body content length per /api/put method to insert datapoints
+	// into opentsdb.
+	// This value is optional, and if it is not set, client.defaultMaxPutPointsNum
+	// will be used in the opentsdb client.
+	MaxContentLength int
+}
+
+type Health struct {
+	Status  string         `json:"status,omitempty"`
+	Details map[string]any `json:"details,omitempty"`
+}
+
+// New initializes a new instance of Opentsdb with provided configuration.
+func New(config *Config) *Client {
+	return &Client{config: *config}
+}
+
+func (c *Client) UseLogger(logger interface{}) {
+	if l, ok := logger.(Logger); ok {
+		c.logger = l
+	}
+}
+
+func (c *Client) UseMetrics(metrics interface{}) {
+	if m, ok := metrics.(Metrics); ok {
+		c.metrics = m
+	}
+}
+
+func (c *Client) UseTracer(tracer any) {
+	if tracer, ok := tracer.(trace.Tracer); ok {
+		c.tracer = tracer
+	}
+}
+
+// Connect initializes an HTTP client for OpenTSDB using the provided configuration.
+// If the configuration is invalid or the endpoint is unreachable, an error is logged.
+func (c *Client) Connect() {
+	span := c.addTrace(context.Background(), "Connect")
+
+	status := statusFailed
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "Connect", &status, &message, span)
+
+	// Set default values for optional configuration fields.
+	c.initializeClient()
+
+	// Initialize the OpenTSDB client with the given configuration.
+	c.endpoint = fmt.Sprintf("http://%s", c.config.Host)
+
+	c.logger.Logf("Connection Successful")
+
+	status = statusSuccess
+	message = fmt.Sprintf("connected to %s", c.endpoint)
+}
+
+func (c *Client) PutDataPoints(ctx context.Context, datas any, queryParam string, resp any) error {
+	span := c.addTrace(ctx, "Put")
+
+	status := statusFailed
+
+	message := "Put request failed"
+
+	defer sendOperationStats(c.logger, time.Now(), "Put", &status, &message, span)
+
+	putResp, ok := resp.(*PutResponse)
+	if !ok {
+		return errors.New("invalid response type. Must be *PutResponse")
+	}
+
+	datapoints, ok := datas.([]DataPoint)
+	if !ok {
+		return errors.New("invalid response type. Must be []DataPoint")
+	}
+
+	err := validateDataPoint(datapoints)
+	if err != nil {
+		message = fmt.Sprintf("invalid data: %s", err)
+		return err
+	}
+
+	if !isValidPutParam(queryParam) {
+		message = "The given query param is invalid."
+		return errors.New(message)
+	}
+
+	var putEndpoint string
+	if !isEmptyPutParam(queryParam) {
+		putEndpoint = fmt.Sprintf("%s%s?%s", c.endpoint, putPath, queryParam)
+	} else {
+		putEndpoint = fmt.Sprintf("%s%s", c.endpoint, putPath)
+	}
+
+	tempResp, err := c.getResponse(ctx, putEndpoint, datapoints, &message)
+	if err != nil {
+		return err
+	}
+
+	if len(tempResp.Errors) == 0 {
+		status = statusSuccess
+		message = fmt.Sprintf("Put request to url %q processed successfully", putEndpoint)
+		putResp.Success = tempResp.Success
+		putResp.Failed = tempResp.Failed
+		putResp.Errors = tempResp.Errors
```

Comment:
PutResp has only 3 fields

I'm surprised to see this here

Are you trying to do something like

```suggestion
		*putResp = tempResp
```